### PR TITLE
Execute jobs in separate conda environments

### PR DESCRIPTION
**Pull request by jan-janssen**
**Caution:** Executing pyiron_base jobs in separate conda environments introduces additional overhead. When possible it is highly recommended to create one conda environment which combines all dependencies. 

# Feature: 
For specific application it is required to execute parts of a given workflow in separate conda environments. For example in [BAMresearch/NFDI4IngScientificWorkflowRequirements](https://github.com/BAMresearch/NFDI4IngScientificWorkflowRequirements). For these special cases this pull request adds the functionality to `pyiron_base` to execute individual jobs in separate conda environments. Additional conda environments can be created with the `Project()` object using `pr.conda_environment.create()` and can be assigned to individual jobs using `job.server.conda_environment_path = pr.conda_environment.py312`. 

# Example:
```
pr = Project("Test")
pr.conda_environment.create(env_name="py312", env_file="env.yaml")
pr.create_job_class(
    class_name="PythonVersionJob",
    executable_str="python --version",
)
job = pr.create.job.PythonVersionJob(job_name="job_conda_path")
job.server.conda_environment_path = pr.conda_environment.py312
job.run()
```

### CODE REVIEW

**Comment by jan-janssen** [comment]
It would be great if we would have python bindings for the creation of conda environments - https://github.com/pyiron/conda_subprocess/issues/4

**Review by niklassiemer — COMMENTED** [review]
I see the value of the addition in principle, however, tab completion would only show `create` and the getitem is hidden. Just using pr.conda_environment would only inform that this is a property. 
Therefore, I would add a get method and provide a reasonable __repr__ such that things are clear from the user perspective.

**Comment by jan-janssen** [comment]
> I see the value of the addition in principle, however, tab completion would only show `create` and the getitem is hidden. Just using pr.conda_environment would only inform that this is a property. Therefore, I would add a get method and provide a reasonable **repr** such that things are clear from the user perspective.

I agree a `__repr__` function should be added for visualisation. Still I do not understand your concerns about the tab completion, as I overload the `__dir__` method. I just missed to add the `create` method in the `__dir__` function.

File: pyiron_base/project/condaenv.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        item_dict = {os.path.basename(path): path for path in list_all_known_prefixes()}
+        if item in item_dict.keys():
+            return item_dict[item]
+        else:
+            raise AttributeError()
+
+    @staticmethod
+    def create(env_name, env_file):
+        subprocess.check_output(
+            ["mamba", "env", "create", "-n", env_name, "-f", env_file, "-y"],
```
Comment (niklassiemer) [inline pyiron_base/project/condaenv.py]: mamba or conda?
```suggestion
    def create(env_name, env_file, use_mamba=True):
        exe = "mamba" if use_mamba else "conda"
        subprocess.check_output(
            [exe, "env", "create", "-n", env_name, "-f", env_file, "-y"],
```

File: pyiron_base/project/condaenv.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    @staticmethod
+    def _list_all_known_prefixes_dict():
+        return {os.path.basename(path): path for path in list_all_known_prefixes()}
+
+    def __getattr__(self, item):
+        item_dict = {os.path.basename(path): path for path in list_all_known_prefixes()}
+        if item in item_dict.keys():
+            return item_dict[item]
+        else:
+            raise AttributeError()
```
Comment (niklassiemer) [inline pyiron_base/project/condaenv.py]:
```suggestion
            raise AttributeError(f"Unknown conda environment {item}. Use one of {self._list_all_known_prefixes_dict()} or create a new one.")
```

File: pyiron_base/project/generic.py [hunk 3]
```diff
@@ -224,6 +224,12 @@ def size(self):
         )
         return self._size_conversion(size)
 
+    @property
+    def conda_environment(self):
+        from pyiron_base.project.condaenv import CondaEnvironment
+
```
Comment (niklassiemer) [inline pyiron_base/project/generic.py]:
```suggestion
    def conda_environment(self):
        try:
            from pyiron_base.project.condaenv import CondaEnvironment
        except ImportError as e:
            raise ImportError("You need to have the conda python package installed to access conda environments.") from None

```

**Review by niklassiemer — COMMENTED** [review]
Yes, sorry I misread `getattr` and `getitem` and was confused!

**Comment by niklassiemer** [comment]
I did understand correctly, that you only use the other conda environment for starting the shell script running the actual computation? Or is it required that pyiron(_base) is also installed in the other conda environment?

File: tests/flex/test_executablecontainer_conda.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+
+@unittest.skipIf(skip_conda_test, "conda is not available, so the conda environment related tests are skipped.")
+class TestExecutableContainerConda(TestWithProject):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        with open("env.yaml", "w") as f:
+            f.writelines(conda_env)
+        cls.project.conda_environment.create(env_name="py312", env_file="env.yaml")
```
Comment (jan-janssen) [inline tests/flex/test_executablecontainer_conda.py:28]: The creation of conda environments fails on windows, so the windows tests are skipped for now.

**Comment by jan-janssen** [comment]
> I did understand correctly, that you only use the other conda environment for starting the shell script running the actual computation? Or is it required that pyiron(_base) is also installed in the other conda environment?

That is correct, only the external executable is executed in the different environment. This allows the external environments to be small and simplifies the integration on the pyiron side. Calling python functions in other conda environments is currently not supported.

**Pull request merged**